sklearn/linear_model/tests/test_logistic.py

```diff
@@ -2374,3 +2374,47 @@ def test_multi_class_deprecated():
     lrCV = LogisticRegressionCV(multi_class="multinomial")
     with pytest.warns(FutureWarning, match=msg):
         lrCV.fit(X, y)
+
+
+def test_newton_cholesky_fallback_to_lbfgs(global_random_seed):
+    # Wide data matrix should lead to a rank-deficient Hessian matrix
+    # hence make the Newton-Cholesky solver raise a warning and fallback to
+    # lbfgs.
+    X, y = make_classification(
+        n_samples=10, n_features=20, random_state=global_random_seed
+    )
+    C = 1e30  # very high C to nearly disable regularization
+
+    # Check that LBFGS can converge without any warning on this problem.
+    lr_lbfgs = LogisticRegression(solver="lbfgs", C=C)
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")
+        lr_lbfgs.fit(X, y)
+        n_iter_lbfgs = lr_lbfgs.n_iter_[0]
+
+    assert n_iter_lbfgs >= 1
+
+    # Check that the Newton-Cholesky solver raises a warning and falls back to
+    # LBFGS. This should converge with the same number of iterations as the
+    # above call of lbfgs since the Newton-Cholesky triggers the fallback
+    # before completing the first iteration, for the problem setting at hand.
+    lr_nc = LogisticRegression(solver="newton-cholesky", C=C)
+    with ignore_warnings(category=LinAlgWarning):
+        lr_nc.fit(X, y)
+        n_iter_nc = lr_nc.n_iter_[0]
+
+    assert n_iter_nc == n_iter_lbfgs
+
+    # Trying to fit the same model again with a small iteration budget should
+    # therefore raise a ConvergenceWarning:
+    lr_nc_limited = LogisticRegression(
+        solver="newton-cholesky", C=C, max_iter=n_iter_lbfgs - 1
+    )
+    with ignore_warnings(category=LinAlgWarning):
+        with pytest.warns(ConvergenceWarning, match="lbfgs failed to converge"):
+            lr_nc_limited.fit(X, y)
+            n_iter_nc_limited = lr_nc_limited.n_iter_[0]
+
+    # XXX: Is this a one-off error? Shouln't it be equal to
+    # lr_nc_limited.max_iter instead?
+    assert n_iter_nc_limited == lr_nc_limited.max_iter - 1
```

Comment:
does this need to be resolved?

Reply:
I am not sure if this is a bug or not. We would need to evaluate how other scikit-learn estimator behave in that respect (and maybe do a common test to enforce consistency).

But it's independent of the particular bug this PR fixes.

Reply:
I would remove the comment.
And tbh (what follows is a frank statement of my personal opinion), I don't consider this a bug. We trigger a convergence warning anyway. In this exceptional case, the solver tries not to abandon completely, but to give a meaningful result (the coefficients). I don't really care how to exactly count the numbers of iterations in such a scenario, it's screwed anyway. Don't we have more important things/inconsistencies to fix?

Reply:
Point taken.